proposals/1547_impersonation-proxy-external-certs/README.md

```diff
@@ -0,0 +1,167 @@
+
+#### API Changes
+
+```yaml
+apiVersion: "config.concierge.pinniped.dev/v1alpha1"
+kind: CredentialIssuer
+metadata:
+  name: the-credential-issuer
+spec:
+  impersonationProxy:
+    mode: auto
+    externalEndpoint: impersonation-proxy.example.com
+    service:
+      loadBalancerIP: 1.2.3.4
+    # Proposed API below:
+    tls:
+      # certificateAuthorityData contains a CA bundle. This value is not used by the impersation proxy to serve TLS.
+      # This value will be advertised to clients so that they can perform TLS verification with the impersonation proxy.
+      # Specifying multiple CA certs can assist with CA rotation.
+      # Optional. If not provided, no CA bundle will be advertised for clients.
+      certificateAuthorityData: <ca-bundle>
+
+      # Names a secret of type "kubernetes.io/tls" (https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets)
+      # which must contain both a TLS serving certificate and the private key, and which is in the same namespace.
+      # Eventually, this will support using the "ca.crt" field which is sometimes provided by cert-manager
+      # (https://cert-manager.io/docs/concepts/certificate/), instead of providing certificateAuthorityData above.
+      # Eventually, this serving certificate may be validated against the above externalEndpoint and/or loadBalancerIP.
+      # Required if the tls parent is set.
+      secretName: my-tls-cert
+```
+
+#### Upgrades
+
+* Upgrading an existing impersonation proxy installation currently using a generated cert should continue to work as-is
+  without intervention
+* Upgrading an existing impersonation proxy installation currently using a generated cert should allow easy transfer to
+  an external cert.
+  The impersonation proxy will clean up its own generated certs that are no longer used.
+  This will require manual intervention for at least the following:
+    * Configure the external cert secret (using cert-manager, manually generated certs, or any other mechanism)
+    * Configure the CredentialIssuer with the new tls configuration block
+    * Regenerate and distribute a new kubeconfig for that cluster
+* Installing a new impersonation proxy with an external cert should work without ever generating a cert
+* Switching an impersonation proxy from an external cert to a generated cert should work by performing the following
+  manual interventions:
+    * Remove the new tls configuration block from the CredentialIssuer
+    * Clean up existing external CA/cert secret objects, and prevent their regeneration
+    * Regenerate and distribute a new kubeconfig for that cluster
+
+#### Tests
+
+No new integration tests, since the verification likely requires multiple components not typically installed during
+integration tests
+such as cert-manager and contour.
```

Comment:
We can have integration tests which generate their own CA and TLS certs using golang helpers (we already have these helpers, e.g. for generating Supervisor certs). Although this feature could be used with an ingress system like Contour, we shouldn't need that for tests. It should be possible to add new tests in `test/integration/concierge_impersonation_proxy_test.go`, which already configures/reconfigures the impersonation proxy via the CredentialIssuer CR.

Reply:
Updated to reflect that we *will* write integration tests

Reply:
I think you forgot to update the section called "Tests".